File: verify-strings.cpp
// Comprehensive string key verification for abseil
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <algorithm>
#include <chrono>
#include <vector>
#include <string>
#include <string_view>
#include <random>
#include "absl/container/flat_hash_map.h"
#include "absl/hash/hash.h"

template <typename T>
inline void do_not_optimize(T const& val) {
    asm volatile("" : : "r,m"(val) : "memory");
}

static uint64_t splitmix64(uint64_t& state) {
    state += 0x9e3779b97f4a7c15;
    uint64_t z = state;
    z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
    z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
    return z ^ (z >> 31);
}

constexpr uint64_t KEY_SEED = 0xDEADBEEF12345678;
constexpr int TOTAL_RUNS = 12;
constexpr int WARMUP = 2;
constexpr int MEASURED = TOTAL_RUNS - WARMUP;
static const char hex_chars[] = "0123456789abcdef";

static void gen_hex_key(uint64_t val, char* buf, size_t len) {
    // Fill with hex from splitmix64 values, repeating if needed
    for (size_t i = 0; i < len; i++) {
        buf[i] = hex_chars[(val >> ((i % 16) * 4)) & 0xF];
        if (i % 16 == 15) val = val * 0x517cc1b727220a95 + i; // mix for longer keys
    }
}

static uint64_t median_val(uint64_t* arr, int n) {
    std::sort(arr, arr + n);
    return arr[n / 2];
}

// Test 1: Hash function cost for strings
static void test_hash_cost() {
    printf("=== Test 1: Hash function cost (16-byte strings) ===\n");
    constexpr size_t N = 10000000;
    char keys[1024][16];
    uint64_t ks = KEY_SEED;
    for (int i = 0; i < 1024; i++) {
        uint64_t v = splitmix64(ks);
        for (int j = 15; j >= 0; j--) { keys[i][j] = hex_chars[v & 0xF]; v >>= 4; }
    }

    absl::Hash<std::string_view> hasher;
    uint64_t best = UINT64_MAX;
    for (int r = 0; r < 10; r++) {
        auto start = std::chrono::steady_clock::now();
        uint64_t acc = 0;
        for (size_t i = 0; i < N; i++) {
            acc += hasher(std::string_view(keys[i & 1023], 16));
        }
        auto end = std::chrono::steady_clock::now();
        do_not_optimize(acc);
        uint64_t us = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
        if (us < best) best = us;
    }
    printf("  absl::Hash<string_view>(16B): %lu us (%.1f ns/hash)\n\n", best, best * 1000.0 / N);
}

// Test 2: Variable key lengths (shuffled lookup, 1M at 50% load)
static void test_variable_lengths() {
    printf("=== Test 2: Variable key lengths (1M, 50%% load, shuffled) ===\n");
    constexpr size_t n = 1048576;
    constexpr size_t fill = n / 2;
    constexpr size_t key_lens[] = {8, 16, 32, 64, 128, 256};

    for (size_t klen : key_lens) {
        std::vector<char> key_buf(fill * klen);
        uint64_t ks = KEY_SEED;
        for (size_t i = 0; i < fill; i++) {
            gen_hex_key(splitmix64(ks), &key_buf[i * klen], klen);
        }

        std::vector<std::string_view> keys(fill);
        for (size_t i = 0; i < fill; i++)
            keys[i] = std::string_view(&key_buf[i * klen], klen);

        // Shuffle
        std::vector<size_t> order(fill);
        for (size_t i = 0; i < fill; i++) order[i] = i;
        std::mt19937_64 rng(42);
        for (size_t i = fill - 1; i > 0; i--) { size_t j = rng() % (i + 1); std::swap(order[i], order[j]); }

        uint64_t times[MEASURED];
        for (int r = 0; r < TOTAL_RUNS; r++) {
            absl::flat_hash_map<std::string_view, uint64_t> map;
            map.reserve(n);
            for (size_t i = 0; i < fill; i++) map.emplace(keys[i], i);

            auto start = std::chrono::steady_clock::now();
            for (size_t i = 0; i < fill; i++) {
                auto it = map.find(keys[order[i]]);
                do_not_optimize(it->second);
            }
            auto end = std::chrono::steady_clock::now();
            if (r >= WARMUP)
                times[r - WARMUP] = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
        }
        printf("  klen=%3zu: %lu us\n", klen, median_val(times, MEASURED));
    }
    printf("\n");
}

// Test 3: Multiple table sizes at 50% load, shuffled, 16-byte keys
static void test_sizes() {
    printf("=== Test 3: Table sizes (50%% load, shuffled, 16B keys) ===\n");
    constexpr size_t sizes[] = {100000, 500000, 1000000, 2000000, 4000000};

    for (size_t n : sizes) {
        size_t fill = n / 2;
        std::vector<char> key_buf(fill * 16);
        uint64_t ks = KEY_SEED;
        for (size_t i = 0; i < fill; i++) {
            uint64_t v = splitmix64(ks);
            for (int j = 15; j >= 0; j--) { key_buf[i * 16 + j] = hex_chars[v & 0xF]; v >>= 4; }
        }

        std::vector<std::string_view> keys(fill);
        for (size_t i = 0; i < fill; i++)
            keys[i] = std::string_view(&key_buf[i * 16], 16);

        std::vector<size_t> order(fill);
        for (size_t i = 0; i < fill; i++) order[i] = i;
        std::mt19937_64 rng(42);
        for (size_t i = fill - 1; i > 0; i--) { size_t j = rng() % (i + 1); std::swap(order[i], order[j]); }

        uint64_t times[MEASURED];
        for (int r = 0; r < TOTAL_RUNS; r++) {
            absl::flat_hash_map<std::string_view, uint64_t> map;
            map.reserve(n);
            for (size_t i = 0; i < fill; i++) map.emplace(keys[i], i);

            auto start = std::chrono::steady_clock::now();
            for (size_t i = 0; i < fill; i++) {
                auto it = map.find(keys[order[i]]);
                do_not_optimize(it->second);
            }
            auto end = std::chrono::steady_clock::now();
            if (r >= WARMUP)
                times[r - WARMUP] = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
        }
        printf("  n=%7zu fill=%7zu: %lu us\n", n, fill, median_val(times, MEASURED));
    }
    printf("\n");
}

// Test 4: std::string (owning) vs string_view at 1M 50%
static void test_owning_strings() {
    printf("=== Test 4: std::string (owning) vs string_view (1M, 50%%, shuffled) ===\n");
    constexpr size_t n = 1048576;
    constexpr size_t fill = n / 2;

    std::vector<char> key_buf(fill * 16);
    uint64_t ks = KEY_SEED;
    for (size_t i = 0; i < fill; i++) {
        uint64_t v = splitmix64(ks);
        for (int j = 15; j >= 0; j--) { key_buf[i * 16 + j] = hex_chars[v & 0xF]; v >>= 4; }
    }

    std::vector<std::string> owned_keys(fill);
    std::vector<std::string_view> sv_keys(fill);
    for (size_t i = 0; i < fill; i++) {
        owned_keys[i] = std::string(&key_buf[i * 16], 16);
        sv_keys[i] = std::string_view(&key_buf[i * 16], 16);
    }

    std::vector<size_t> order(fill);
    for (size_t i = 0; i < fill; i++) order[i] = i;
    std::mt19937_64 rng(42);
    for (size_t i = fill - 1; i > 0; i--) { size_t j = rng() % (i + 1); std::swap(order[i], order[j]); }

    // string_view map
    uint64_t sv_times[MEASURED];
    for (int r = 0; r < TOTAL_RUNS; r++) {
        absl::flat_hash_map<std::string_view, uint64_t> map;
        map.reserve(n);
        for (size_t i = 0; i < fill; i++) map.emplace(sv_keys[i], i);

        auto start = std::chrono::steady_clock::now();
        for (size_t i = 0; i < fill; i++) {
            auto it = map.find(sv_keys[order[i]]);
            do_not_optimize(it->second);
        }
        auto end = std::chrono::steady_clock::now();
        if (r >= WARMUP)
            sv_times[r - WARMUP] = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
    }

    // std::string map
    uint64_t str_times[MEASURED];
    for (int r = 0; r < TOTAL_RUNS; r++) {
        absl::flat_hash_map<std::string, uint64_t> map;
        map.reserve(n);
        for (size_t i = 0; i < fill; i++) map.emplace(owned_keys[i], i);

        auto start = std::chrono::steady_clock::now();
        for (size_t i = 0; i < fill; i++) {
            auto it = map.find(owned_keys[order[i]]);
            do_not_optimize(it->second);
        }
        auto end = std::chrono::steady_clock::now();
        if (r >= WARMUP)
            str_times[r - WARMUP] = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
    }

    printf("  string_view: %lu us\n", median_val(sv_times, MEASURED));
    printf("  std::string: %lu us\n", median_val(str_times, MEASURED));
    printf("  ratio (string/sv): %.2f\n\n", (double)median_val(str_times, MEASURED) / median_val(sv_times, MEASURED));
}

int main() {
    test_hash_cost();
    test_variable_lengths();
    test_sizes();
    test_owning_strings();
    return 0;
}

File: src/verify-strings.zig
//! String key verification: variable lengths, sizes, correctness
const std = @import("std");
const StringElasticHash = @import("string_hybrid.zig").StringElasticHash;
const Wyhash = std.hash.Wyhash;
const linux = std.os.linux;
const sort = std.sort;

const KEY_SEED: u64 = 0xDEADBEEF12345678;
const TOTAL_RUNS: usize = 12;
const WARMUP: usize = 2;
const MEASURED: usize = TOTAL_RUNS - WARMUP;

fn splitmix64(state: *u64) u64 {
    state.* +%= 0x9e3779b97f4a7c15;
    var z = state.*;
    z = (z ^ (z >> 30)) *% 0xbf58476d1ce4e5b9;
    z = (z ^ (z >> 27)) *% 0x94d049bb133111eb;
    return z ^ (z >> 31);
}

const hex_chars = "0123456789abcdef";

fn genHexKey(val: u64, buf: []u8) void {
    var v = val;
    for (buf, 0..) |*b, i| {
        b.* = hex_chars[@as(usize, @intCast((v >> @as(u6, @intCast((i % 16) * 4))) & 0xF))];
        if (i % 16 == 15) v = v *% 0x517cc1b727220a95 +% i;
    }
}

fn now() linux.timespec {
    var ts: linux.timespec = undefined;
    _ = linux.clock_gettime(.MONOTONIC, &ts);
    return ts;
}

fn usElapsed(start: linux.timespec, end: linux.timespec) u64 {
    const s: u64 = @intCast(end.sec - start.sec);
    const ns_diff: i64 = @as(i64, @intCast(end.nsec)) - @as(i64, @intCast(start.nsec));
    return (s * 1_000_000_000 + @as(u64, @intCast(ns_diff))) / 1_000;
}

fn median(arr: *[MEASURED]u64) u64 {
    sort.insertion(u64, arr, {}, sort.asc(u64));
    return arr[MEASURED / 2];
}

fn benchShuffled(allocator: std.mem.Allocator, n: usize, fill: usize, klen: usize) u64 {
    const key_buf = allocator.alloc(u8, fill * klen) catch @panic("alloc");
    defer allocator.free(key_buf);

    var ks: u64 = KEY_SEED;
    for (0..fill) |i| {
        genHexKey(splitmix64(&ks), key_buf[i * klen .. (i + 1) * klen]);
    }

    const shuffle_order = allocator.alloc(usize, fill) catch @panic("alloc");
    defer allocator.free(shuffle_order);
    for (0..fill) |i| shuffle_order[i] = i;
    var rng: u64 = 42;
    var idx: usize = fill;
    while (idx > 1) { idx -= 1; const j = splitmix64(&rng) % (idx + 1); const tmp = shuffle_order[idx]; shuffle_order[idx] = shuffle_order[j]; shuffle_order[j] = tmp; }

    var times: [MEASURED]u64 = undefined;

    for (0..TOTAL_RUNS) |r| {
        var map = StringElasticHash.init(allocator, n) catch @panic("init");
        defer map.deinit();
        for (0..fill) |i| {
            map.insert(key_buf[i * klen .. (i + 1) * klen], i);
        }

        const start = now();
        for (0..fill) |i| {
            const ki = shuffle_order[i];
            std.mem.doNotOptimizeAway(map.get(key_buf[ki * klen .. (ki + 1) * klen]));
        }
        const end = now();
        if (r >= WARMUP) times[r - WARMUP] = usElapsed(start, end);
    }

    return median(&times);
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    // Test 1: Hash cost
    {
        std.debug.print("=== Test 1: Wyhash cost (16-byte strings) ===\n", .{});
        var keys: [1024][16]u8 = undefined;
        var ks: u64 = KEY_SEED;
        for (&keys) |*k| {
            const v = splitmix64(&ks);
            var val = v;
            comptime var i: usize = 16;
            inline while (i > 0) { i -= 1; k[i] = hex_chars[@as(usize, @intCast(val & 0xF))]; val >>= 4; }
        }
        var best: u64 = std.math.maxInt(u64);
        for (0..10) |_| {
            const start = now();
            var acc: u64 = 0;
            for (0..10_000_000) |i| {
                acc +%= Wyhash.hash(0, &keys[i & 1023]);
            }
            const end = now();
            std.mem.doNotOptimizeAway(acc);
            const us = usElapsed(start, end);
            if (us < best) best = us;
        }
        std.debug.print("  Wyhash(16B): {d} us ({d:.1} ns/hash)\n\n", .{ best, @as(f64, @floatFromInt(best)) * 1000.0 / 10_000_000 });
    }

    // Test 2: Variable key lengths (1M, 50% load, shuffled)
    {
        std.debug.print("=== Test 2: Variable key lengths (1M, 50% load, shuffled) ===\n", .{});
        const klens = [_]usize{ 8, 16, 32, 64, 128, 256 };
        for (klens) |klen| {
            const us = benchShuffled(allocator, 1_048_576, 524_288, klen);
            std.debug.print("  klen={d:>3}: {d} us\n", .{ klen, us });
        }
        std.debug.print("\n", .{});
    }

    // Test 3: Multiple sizes at 50% load, shuffled, 16-byte keys
    {
        std.debug.print("=== Test 3: Table sizes (50% load, shuffled, 16B keys) ===\n", .{});
        const sizes = [_]usize{ 100_000, 500_000, 1_000_000, 2_000_000, 4_000_000 };
        for (sizes) |n| {
            const fill = n / 2;
            const us = benchShuffled(allocator, n, fill, 16);
            std.debug.print("  n={d:>7} fill={d:>7}: {d} us\n", .{ n, fill, us });
        }
        std.debug.print("\n", .{});
    }

    // Test 4: Correctness check (1M, 99%, verify all values)
    {
        std.debug.print("=== Test 4: Correctness check (1M, 99%) ===\n", .{});
        const n: usize = 1_048_576;
        const fill = n * 99 / 100;
        const key_buf = allocator.alloc([16]u8, fill) catch @panic("alloc");
        defer allocator.free(key_buf);

        var ks: u64 = KEY_SEED;
        for (0..fill) |i| {
            const v = splitmix64(&ks);
            var val = v;
            comptime var j: usize = 16;
            inline while (j > 0) { j -= 1; key_buf[i][j] = hex_chars[@as(usize, @intCast(val & 0xF))]; val >>= 4; }
        }

        var map = try StringElasticHash.init(allocator, n);
        defer map.deinit();
        for (0..fill) |i| map.insert(&key_buf[i], i);

        var found: usize = 0;
        var correct: usize = 0;
        for (0..fill) |i| {
            if (map.get(&key_buf[i])) |v| {
                found += 1;
                if (v == i) correct += 1;
            }
        }
        std.debug.print("  Inserted: {d}\n", .{fill});
        std.debug.print("  Found:    {d} ({d:.1}%)\n", .{ found, @as(f64, @floatFromInt(found)) / @as(f64, @floatFromInt(fill)) * 100 });
        std.debug.print("  Correct:  {d} ({d:.1}%)\n\n", .{ correct, @as(f64, @floatFromInt(correct)) / @as(f64, @floatFromInt(fill)) * 100 });
    }

    std.debug.print("DONE\n", .{});
}
